backend/cmm_helpers.ml

```diff
@@ -2526,172 +2566,202 @@ let tuplify_function arity =
             Cop
               ( Calloc mode,
-                [ alloc_closure_header ~mode 5 (dbg ());
+                [ alloc_closure_header ~mode
+                    (header_size + 1 + machtype_stored_size arg_type)
+                    (dbg ());
                   Cconst_symbol (name1 ^ "_" ^ Int.to_string (num + 1), dbg ());
                   alloc_closure_info
-                    ~arity:(curried (arity - num - 1))
-                    ~startenv:3 (dbg ()) ~is_last:true;
-                  Cconst_symbol
-                    (name1 ^ "_" ^ Int.to_string (num + 1) ^ "_app", dbg ());
-                  Cvar arg;
-                  Cvar clos ],
-                dbg () )
-          else
-            Cop
-              ( Calloc mode,
-                [ alloc_closure_header ~mode 4 (dbg ());
-                  Cconst_symbol (name1 ^ "_" ^ Int.to_string (num + 1), dbg ());
-                  alloc_closure_info ~arity:(curried 1) ~startenv:2
-                    ~is_last:true (dbg ());
-                  Cvar arg;
-                  Cvar clos ],
-                dbg () ));
-        fun_codegen_options = [];
-        fun_dbg;
-        fun_poll = Default_poll
-      }
-    ::
-    (if arity <= max_arity_optimized && arity - num > 2
-    then
-      let rec iter i =
-        if i <= arity
-        then
-          let arg = V.create_local (Printf.sprintf "arg%d" i) in
-          (arg, typ_val) :: iter (i + 1)
-        else []
-      in
-      let direct_args = iter (num + 2) in
-      let rec iter i args clos =
-        if i = 0
-        then
-          Cop
-            ( Capply (typ_val, Rc_normal),
-              (get_field_gen Asttypes.Mutable (Cvar clos) 2 (dbg ()) :: args)
-              @ [Cvar clos],
-              dbg () )
-        else
-          let newclos = V.create_local "clos" in
-          Clet
-            ( VP.create newclos,
-              get_field_gen Asttypes.Mutable (Cvar clos) 4 (dbg ()),
-              iter (i - 1)
-                (get_field_gen Asttypes.Mutable (Cvar clos) 3 (dbg ()) :: args)
-                newclos )
-      in
-      let fun_args =
-        List.map
-          (fun (arg, ty) -> VP.create arg, ty)
-          (direct_args @ [clos, typ_val])
-      in
-      let fun_name = name1 ^ "_" ^ Int.to_string (num + 1) ^ "_app" in
-      let fun_dbg = placeholder_fun_dbg ~human_name:fun_name in
-      let cf =
-        Cfunction
-          { fun_name;
-            fun_args;
-            fun_body =
-              iter (num + 1)
-                (List.map (fun (arg, _) -> Cvar arg) direct_args)
-                clos;
-            fun_codegen_options = [];
-            fun_dbg;
-            fun_poll = Default_poll
-          }
-      in
-      cf :: intermediate_curry_functions ~nlocal ~arity (num + 1)
-    else intermediate_curry_functions ~nlocal ~arity (num + 1))
+                    ~arity:(curried (if has_nary then narity - num - 1 else 1))
+                    ~startenv:(3 + machtype_non_scanned_size arg_type)
```

Comment:
I suspect this 3 should say header_size.

Reply:
Well spotted!